### ISSUE TITLE
AQL - comparing  vertex._id  to edge._to not working as expected

### ISSUE THREAD

**Issue by sylvainlap** [arangodb/arangodb#2320]
Hello,

I'm using ArangoDB 3.1.11 from the Docker official image.
This is an AQL issue, using traversal.

I have an AQL request, which is a traversal, and inside this traversal, I want to use the _id of the current vertex. Here is my AQL request:

```
FOR v, e, p IN 1..1 OUTBOUND "items/1064" GRAPH "aquarium"
          FILTER p.edges[*].type ALL == "Shared"
      
        LET userGroupAccess = (
          FOR v2, e2, p2 IN 2..2 OUTBOUND "items/671" GRAPH "aquarium"
            FILTER p2.edges[0].type == "Access"
              AND p2.vertices[1].type == "Usergroup"
              AND p2.edges[1].type == "Access"
              AND p2.edges[1]._to == v._id
            RETURN e2
        )
        
      SORT v.updatedAt DESC LIMIT 0, 100 RETURN { userGroupAccess, v }
```

and here is the result:

```
[
  {
    "userGroupAccess": [],
    "v": {
      "_key": "1013",
      "_id": "items/1013",
      "_rev": "_UkTMII----",
      "type": "Note",
      "data": {
        "name": "New Note",
        "content": "..."
      },
      "createdBy": "329",
      "createdAt": "1487758305436",
      "updatedAt": "1487758305437"
    }
  }
]
```

As you can see, the userGroupAccess is empty. However I know for sure that an edge exists, and userGroupAccess should return one edge. Here is the proof: if I change v._id by the actual _id of the vertex:

```
FOR v, e, p IN 1..1 OUTBOUND "items/1064" GRAPH "aquarium"
          FILTER p.edges[*].type ALL == "Shared"
      
        LET userGroupAccess = (
          FOR v2, e2, p2 IN 2..2 OUTBOUND "items/671" GRAPH "aquarium"
            FILTER p2.edges[0].type == "Access"
              AND p2.vertices[1].type == "Usergroup"
              AND p2.edges[1].type == "Access"
              AND p2.edges[1]._to == "items/1013"
            RETURN e2
        )
        
      SORT v.updatedAt DESC LIMIT 0, 100 RETURN { userGroupAccess, v }
```

Here is the result:

```
[
  {
    "userGroupAccess": [
      {
        "_key": "1277",
        "_id": "connections/1277",
        "_from": "items/1064",
        "_to": "items/1013",
        "_rev": "_UkTOelq---",
        "type": "Access",
        "data": {
          "permissions": "r",
          "foo": "bar"
        },
        "createdBy": "329",
        "createdAt": "1487758459512",
        "updatedAt": "1487758459512"
      }
    ],
    "v": {
      "_key": "1013",
      "_id": "items/1013",
      "_rev": "_UkTMII----",
      "type": "Note",
      "data": {
        "name": "New Note",
        "content": "..."
      },
      "createdBy": "329",
      "createdAt": "1487758305436",
      "updatedAt": "1487758305437"
    }
  }
]
```

Now userGroupAccess is no more empty, the edge is return. The only thing I change is that I replace `v._id` by the value it is expected to be (`"items/1013"`), as you can see in the `v` result of the first query...

I think there is a bug in the LET computation, because both requests should return the same result.

Regards,
Sylvain Lapendry

**Comment by jsteemann** [comment]
I tried the query and found that it works fine when turning off the optimizer rule `optimize-traversals`. With the rule applied, the query does not return the correct result.
There was a bug in the traversal optimization that pulled some of the filter conditions inside the traversal and executed them locally in the traversal (in contrast to executing them after the traversal as done with the optimizer rule disabled). The actual problem was that only some types of filter conditions can be evaluated inside the traversal, and here a condition was pulled into the traversal and executed that was not supported. The bugfix I pushed should address that.
The fix will be included in 3.1.12.
